File: Java/_096_Unique_Binary_Search_Trees/Solution.java
class Solution {
    public int numTrees(int n) {
        int[] f = new int[n + 1];
        f[0] = 1;
        f[1] = 1; 
        for (int i = 2; i < n + 1; ++i)
            for (int k = 1; k < i + 1; ++k)
                f[i] = f[i] + f[k - 1] * f[i - k];
        return f[n];
    }
}

File: Python/_096_Unique_Binary_Search_Trees/solution.py
class Solution:
    def numTrees(self, n):
        """
        :type n: int
        :rtype: int
        """
        # f(i)=f(0)*f(i-1)+...+f(k-1)*f(i-k)+...+f(i-1)*f(0) 
        f = [0 for i in range(n + 1)]
        f[0] = 1 
        f[1] = 1  
        for i in range(2, n + 1):
            for k in range(1, i + 1):
                f[i] = f[i] + f[k - 1] * f[i - k]
        return f[n]
